org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/ContentModelCodeLensParticipant.java

```diff
@@ -62,9 +67,60 @@
  */
 public class ContentModelCodeLensParticipant implements ICodeLensParticipant {
 
+	private final ContentModelManager contentModelManager;
+
+	public ContentModelCodeLensParticipant(ContentModelManager contentModelManager) {
+		this.contentModelManager = contentModelManager;
+	}
+
 	@Override
 	public void doCodeLens(ICodeLensRequest request, List<CodeLens> lenses, CancelChecker cancelChecker) {
-		if (!canSupport(request)) {
+		// List of referenced schema/grammar
+		createReferencedGrammarLenses(request, lenses);
+		// "Bind to grammar/schema..."
+		createBindToGrammarSchemaLenses(request, lenses);
+	}
+
+	private void createReferencedGrammarLenses(ICodeLensRequest request, List<CodeLens> lenses) {
+		if (!canSupportOpenUri(request)) {
+			return;
+		}
+		
+		DOMDocument document = request.getDocument();
+		if (!document.hasGrammar()) {
+			return;
+		}
+		Range range = XMLPositionUtility.createRange(document);
+		Set<ReferencedGrammarInfo> referencedGrammarInfos = contentModelManager.getReferencedGrammarInfos(document);
+		for (ReferencedGrammarInfo info : referencedGrammarInfos) {
+			lenses.add(createReferencedGrammarLens(info, range));
+		}
+	}
+
+	private static boolean canSupportOpenUri(ICodeLensRequest request) {
+		if (!request.isSupportedByClient(CodeLensKind.OpenUri)) {
+			return false;
+		}
+		String uri = request.getDocument().getDocumentURI();
+		return !DOMUtils.isXSD(uri) && !DOMUtils.isDTD(uri);
+	}
+	
+	private CodeLens createReferencedGrammarLens(ReferencedGrammarInfo info, Range range) {
+		Identifier identifier = info.getIdentifier();
+		String publicId = identifier.getPublicId();
+		String systemId = identifier.getSystemId();
+		String uri = publicId != null ? publicId : systemId;
+		String title = uri + " (" + identifier.getKind() + ")";
+		String grammarURI = info.isInCache() ? info.getGrammarCacheInfo().getCachedResolvedUri() : null;
+		if (grammarURI == null) {
+			grammarURI = info.getResolvedURIInfo().getResolvedURI();
+		}
+		Command command = new Command(title, OPEN_URI, Arrays.asList(grammarURI));

```

Comment:
It might be nice to pass a argument to the command to tell the client whether the document should be opened in read-only mode. If the document is in the cache, if would be nice to prevent the user from writing to it. If this gets too complex, feel free to ignore.

Reply:
To be consistent with hyperlink (ex : from  xsi:noNamepaceSchemaLaocation value) I would like to avoid opening the xsd in read-only mode.